eth/consensus/pow.py

```diff
@@ -30,23 +31,33 @@
 
 
 # Type annotation here is to ensure we don't accidentally use strings instead of bytes.
-cache_seeds = [b'\x00' * 32]  # type: List[bytes]
+cache_seeds = {0: b'\x00' * 32}  # type: Dict[int, bytes]
 cache_by_seed = OrderedDict()  # type: OrderedDict[bytes, bytearray]
 CACHE_MAX_ITEMS = 10
 
 
 def get_cache(block_number: int) -> bytes:
-    while len(cache_seeds) <= block_number // EPOCH_LENGTH:
-        cache_seeds.append(keccak(cache_seeds[-1]))
-    seed = cache_seeds[block_number // EPOCH_LENGTH]
+    # Generate the seed for the cache if it's missing, aim for thread-safety
+    epoch_index = block_number // EPOCH_LENGTH
+    while epoch_index not in cache_seeds:
+        highest_index = max(cache_seeds.keys())
+        cache_seeds[highest_index + 1] = keccak(cache_seeds[highest_index])
+    seed = cache_seeds[epoch_index]
+
+    # Get the cache if already generated, marking it as recently used
```

Comment:
Can we drop the `cache_by_seed` data structure (and the need to manage it's size) by just decorating `get_cache` with `@functools.lru_cache(max_size=10)`?  I feel like I must be missing something that prevents that.

Reply:
We want to reuse one cache entry for all blocks in the same epoch.

Maybe you mean that we could split this into two functions, so that `get_cache(block_number)` calls `get_cached_pow_cache(seed)`, then `get_cached_pow_cache()` could have a simple lru cache.

Reply:
Actually, I'm super glad you asked because *most* of this code is unnecessary. We're not even using the seed for anything! I went back in history to see why, and don't see a good reason: https://github.com/ethereum/py-evm/commit/78ab70915e8cdd500ba8dffd710a7465ff1e8bd9

Looks like it was just a bug. This is about to become two 2-line functions :man_facepalming:

Reply:
Well, I made it a lot simpler, like just:

```py
def get_cache(block_number: int) -> bytes:                                                                             
    epoch_index = block_number // EPOCH_LENGTH                                                                         
    return _get_cached_pow_cache(epoch_index)                                                                          
                                                                                                                       
                                                                                                                       
@functools.lru_cache(maxsize=CACHE_MAX_ITEMS)                                                                          
def _get_cached_pow_cache(epoch_index: int) -> bytes:                                                                  
    # simulate requesting mkcache by block number: multiply index by epoch length                                      
    return mkcache_bytes(epoch_index * EPOCH_LENGTH)  
```

That caused the tests to run 70% slower, which was a bit of a surprise to me. We didn't have to do *any* keccak hashing. But I guess the lru_cache overhead is higher than I thought and/or is especially ineffective in a heavy threaded environment.

Reply:
Seemed like too high a perf penalty to pay. I did remove the pointless hashing, but kept the manual caching, plus a note explaining why.